dotnet/src/Connectors/Connectors.OpenAI/Core/OpenAIJsonSchemaTransformer.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Linq;
+using System.Text.Json.Nodes;
+using JsonSchemaMapper;
+
+namespace Microsoft.SemanticKernel.Connectors.OpenAI;
+
+/// <summary>
+/// JSON Schema transformer to apply OpenAI conditions for structured outputs.
+/// <para>
+/// - "additionalProperties" property must always be set to <see langword="false"/> in objects.
+/// More information here: <see href="https://platform.openai.com/docs/guides/structured-outputs/additionalproperties-false-must-always-be-set-in-objects"/>.
+/// </para>
+/// <para>
+/// - All fields must be "required".
+/// More information here: <see href="https://platform.openai.com/docs/guides/structured-outputs/all-fields-must-be-required"/>.
+/// </para>
+/// </summary>
+internal static class OpenAIJsonSchemaTransformer
+{
+    private const string AdditionalPropertiesPropertyName = "additionalProperties";
+    private const string TypePropertyName = "type";
+    private const string ObjectValueName = "object";
+    private const string PropertiesPropertyName = "properties";
+    private const string RequiredPropertyName = "required";
+
+    internal static JsonNode Transform(JsonSchemaGenerationContext context, JsonNode schema)
+    {
+        // Transform schema if node is object only.
+        if (schema is JsonObject jsonSchemaObject &&
+            jsonSchemaObject.TryGetPropertyValue(TypePropertyName, out var typeProperty) &&
+            typeProperty is not null)
+        {
+            var type = typeProperty.GetValue<string>();
```

Comment:
What about the cases were `type` is an array of strings?

Reply:
I think it's applicable to `object` only, because if it's array of strings, it contains just values, there are no `required` and `additionalProperties` properties which could be applicable. I provided an example where `required` and `additionalProperties` are provided when type is `object`, but not when type is `array`, and it works correctly:
![image](https://github.com/user-attachments/assets/1adf5e87-0997-4910-9d40-9203acf3b26a)

Reply:
> I think it's applicable to object only, because if it's array of strings, it contains just values, there are no required and additionalProperties properties which could be applicable.

I don't believe that's true, for example consider the schema generated for nullable structs:

```C#
Console.WriteLine(JsonSerializerOptions.Default.GetJsonSchema(typeof(MyPoco?)));
//{
//    "type": ["object", "null"],
//    "properties": {
//        "Id": { "type": "string"}
//     },
//    "additionalProperties": false
//}

[JsonUnmappedMemberHandling(JsonUnmappedMemberHandling.Disallow)]
record struct MyPoco(string Id);
```

Reply:
@eiriktsarpalis Just a small reminder, that this schema is constructed for OpenAI API, so they may handle it differently and not all of JSON schema features are supported. For example, it doesn't look like they support multiple type values in schema. I tried to use your example and I'm getting 400 Bad Request with message: 
```
schema must be a JSON Schema of 'type: "object"', got 'type: "['object', 'null']"'
```

Root object must always be `"type": "object"`.

Here is more information about it:
https://platform.openai.com/docs/guides/structured-outputs/supported-schemas

On our side, I don't see a case when user would pass nullable type as response format, but we can add additional validation to make sure that user doesn't pass nullable type.

Reply:
> Root object must always be "type": "object".

I'm aware, but keep in mind that this could occur as a nested schema, for example:
```C#
record struct RootValue(NestedValue? Nested);
record struct NestedValue(string Value);
```
Yielding the schema
```json
{
  "type": "object",
  "properties": {
    "Nested": {
      "type": ["object","null"],
      "properties": {
        "Value": {
          "type": "string"
        }
      }
    }
  }
}
```
And because the `TransformSchemaNode` is being invoked for every schema node within the document (using depth-first traversal), this particular line of code is going to throw an exception.

> Here is more information about it:
> https://platform.openai.com/docs/guides/structured-outputs/supported-schemas

Yep. You might want to add additional testing for the following configurations that are known to produce JSON schema not accepted by Open AI such as:

1. `DateTime`, `DateTimeOffset`, `DateOnly`, `TimeSpan`, `TimeOnly`, `Uri` that emit the unsupported `format` keyword.
2. `char` which emits `minLength` and `maxLength`,
3. `JsonSerializerOptions.NumberHandling` set to `JsonNumberHandling.AllowReadingFromString` since numeric types then emit the `pattern` keyword.

These would need to stripped from the generated schema using the `TransformSchemaNode` callback. There is some discussion about this in https://github.com/dotnet/runtime/issues/107501. FYI @stephentoub since it seems many users end up doing the same thing to render an OpenAI conforming JSON schema. 

> On our side, I don't see a case when user would pass nullable type as response format,

Does this matter if the APIs permit it in principle?

Reply:
> I'm aware, but keep in mind that this could occur as a nested schema

That makes sense, I will ensure this case is supported.

> You might want to add additional testing for the following configurations that are known to produce JSON schema not accepted by Open AI

> These would need to stripped from the generated schema using the TransformSchemaNode callback.

Thanks for these cases, it's really helpful. If it's not supported by OpenAI, I'm not sure if we need to transform it to something that is supported, strip it or pass it as it is and return an error, so user will be aware about the issue. I think the last option should work for us, because:
- We can't really transform it to something that is supported by OpenAI, because there may be some problems during deserialization later.
- Stripping it silently from JSON schema will simply ignore it and users wouldn't understand why this is happening and why they don't receive specific property in their responses.

It looks like returning an error is the best approach here, because users will be aware about the limitations, which will allow them to define their types properly. The only question is - do we want to perform a validation on Semantic Kernel side or send an invalid schema to OpenAI and return a received error to user. Performing validation on Semantic Kernel side would allow to return well-formatted response with detailed explanation which property is problematic and avoid sending invalid requests to OpenAI. But on the other hand, if OpenAI will gradually enable schema features that are unsupported at the moment, we will have to update the validation logic every time.

Reply:
@eiriktsarpalis Thanks for your review and providing the edge cases! 

I've added a logic to support cases when `type` property has an array of strings value. Now looking at your original comment `What about the cases were type is an array of strings?` I think I misunderstood it, because I thought that you meant a case when `"type": "array"`, and the items inside that array are strings, but now I understand that you meant cases like `"type": "["object", "null"]"` :) 

I've added support for it and also added a logic to handle cases when `Type` is a nullable struct. I've also added related unit-tests.

In terms of cases, that are not supported by OpenAI, like `format`, `minLength`, `maxLength` etc, my proposal is to rely on OpenAI response and return it back to the user as for now. In case there will be some problems with this approach, we can change the behavior by stripping unsupported properties or providing well-formatted message which will describe a problem in more detail. This functionality is experimental, so it would be nice to get early feedback from users to understand if current version would be enough, or additional handling will be needed.

Please let me know if that sounds acceptable. Thank you!

Reply:
I think it's reasonable. We should try to document recommendations to users so that they model their types in a way that makes them compatible with OpenAI restrictions. I have captured some of these recommendations here:

https://github.com/dotnet/runtime/issues/107501#issuecomment-2337883802